src/serac/numerics/functional/detail/hexahedron_H1.inl

```diff
@@ -204,85 +228,195 @@ struct finite_element<mfem::Geometry::CUBE, H1<p, c> > {
       gradient(i, 1) = contract<0, 1>(A22, B);
       gradient(i, 2) = contract<0, 1>(A20, G);
     }
+#else
+
+    RAJA_TEAM_SHARED tensor<double, c, q, q, q> value;
+    RAJA_TEAM_SHARED tensor<double, c, dim, q, q, q> gradient;
+    constexpr auto                                   B = calculate_B<apply_weights, q>();
+    constexpr auto                                   G = calculate_G<apply_weights, q>();
+    for (int i = 0; i < c; i++) {
+      RAJA_TEAM_SHARED decltype(deduce_contract_return_type<2, 1>(X[i], B)) A10;
+      RAJA_TEAM_SHARED decltype(deduce_contract_return_type<2, 1>(X[i], G)) A11;
+      RAJA_TEAM_SHARED decltype(deduce_contract_return_type<1, 1>(A10, B))  A20;
+      RAJA_TEAM_SHARED decltype(deduce_contract_return_type<1, 1>(A11, B))  A21;
+      RAJA_TEAM_SHARED decltype(deduce_contract_return_type<1, 1>(A10, G))  A22;
+
+      RAJA::loop<threads_x>(ctx, x_range, [&](int tid) {
+        int qx = tid % BLOCK_X;
+        int qy = (tid / BLOCK_X) % BLOCK_Y;
+        int qz = (tid / (BLOCK_X * BLOCK_Y)) % BLOCK_Z;
+
+        // Perform actual contractions
+        contract<2, 1>(X[i], B, &A10, qx, qy, qz);
+
+        ctx.teamSync();
+        contract<2, 1>(X[i], G, &A11, qx, qy, qz);
+
+        ctx.teamSync();
+        contract<1, 1>(A10, B, &A20, qx, qy, qz);
+
+        ctx.teamSync();
+        contract<1, 1>(A11, B, &A21, qx, qy, qz);
+
+        ctx.teamSync();
+        contract<1, 1>(A10, G, &A22, qx, qy, qz);

```

Comment:
I think this should be refactored, there's a lot of implementation details leaking through the interface here (e.g. passing qx, qy, qz into the contract function, synchronization, etc).

Hardcoding BLOCK_X, BLOCK_Y and BLOCK_Z doesn't make sense to me, those values depend on the quadrature rule and polynomial order, and are different for each contraction. I believe I sent you some code for this earlier, maybe we can discuss this in more detail.